service/storage_service.cc

```diff
@@ -561,6 +566,66 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
 
         bool res;
         switch (node.rs->ring.value().state) {
+            case ring_slice::replication_state::commit_cdc_generation: {
+                // make sure all nodes know about new topology and have the new CDC generation data
+                // (we require all nodes to be alive for topo change for now)
+                std::tie(node, res) = co_await exec_global_command(std::move(node), raft_topology_cmd{raft_topology_cmd::command::barrier}, false, replaced_node);
```

Comment:
> It's basically an availability check - we shouldn't attempt to pick the generation timestamp if some nodes are not currently available, or we increase the risk that the timestamp doesn't arrive at some nodes in time (causing them to coordinate writes to wrong streams).

Bootstraping is a long process and nodes may become unavailable at any point in time. We should be able to recover.
It is fine to check with failure detection module before starting a topo op that all nodes are alive and refuse to proceed otherwise but inserting barriers in the middle will not help.

> 
> Also this barrier guarantees that the data is present on all nodes before we pick the timestamp. If the generation data is large, replicating it to all nodes might take some significant amount of time. On the other hand, once we pick the timestamp, we want to propagate it ASAP. So it's good to make sure that the generation data is already propagated fully before picking the timestamp, so the timestamp itself can be propagated quickly, without waiting on the data.

How ASAP is ASAP? What are the consequences of not doing it ASAP? Because there is no any timing guaranties how fast it will be propagated and thus the correctness should not depend on it.

Reply:
> Bootstraping is a long process and nodes may become unavailable at any point in time. We should be able to recover.

Bootstrapping in general yes because it includes streaming, but that doesn't matter here. The CDC generation switch is 1. create a timestamp 2. propagate it to the cluster and that's it. We can fail later during streaming, it's fine, CDC doesn't care.

> How ASAP is ASAP? What are the consequences of not doing it ASAP?

Writes to CDC-enabled tables will end up in the wrong CDC streams, causing CDC clients to miss them.
ASAP = before any node's clocks crosses the picked timestamp. We pick now() + 2*ring_delay.

> Because there is no any timing guaranties how fast it will be propagated and thus the correctness should not depend on it.

Unfortunately whoever designed the original bootstrap procedure for Cassandra decided to ignore this fact and introduce a thing called ring_delay. When we designed CDC generations we continued to depend on ring_delay.

Anyway that's why I left the big FIXME suggesting a different 2-phase algorithm for switching generations. (I actually proposed this algorithm when we initially designed CDC, but it depends on having a fault-tolerant coordinator which we only have now)

Reply:
> > Bootstraping is a long process and nodes may become unavailable at any point in time. We should be able to recover.
> 
> Bootstrapping in general yes because it includes streaming, but that doesn't matter here. The CDC generation switch is 1. create a timestamp 2. propagate it to the cluster and that's it. We can fail later during streaming, it's fine, CDC doesn't care.

So we do not need to undo the operation and may just move to the old token metadata if this fails?

> 
> > How ASAP is ASAP? What are the consequences of not doing it ASAP?
> 
> Writes to CDC-enabled tables will end up in the wrong CDC streams, causing CDC clients to miss them. ASAP = before any node's clocks crosses the picked timestamp. We pick now() + 2*ring_delay.

Can we use fencing to reject those writes and make original request fail?

> 
> > Because there is no any timing guaranties how fast it will be propagated and thus the correctness should not depend on it.
> 
> Unfortunately whoever designed the original bootstrap procedure for Cassandra decided to ignore this fact and introduce a thing called ring_delay. When we designed CDC generations we continued to depend on ring_delay.
> 

But we try to amend this with the topology change coordinator + fencing.

Reply:
> So we do not need to undo the operation and may just move to the old token metadata if this fails?

Yes, the new generation is calculated to work well with both old and new token ring. It's just not optimal for the new token ring - contains too many streams. It's preferable to switch the generation back to a smaller one after we abort, but it's optional.

> Can we use fencing to reject those writes and make original request fail?

Yes, but it doesn't help, because a coordinator may finish a write using replicas that did not get the new topology/generation so they won't fence.

Reply:
> > Can we use fencing to reject those writes and make original request fail?
> 
> Yes, but it doesn't help, because a coordinator may finish a write using replicas that did not get the new topology/generation so they won't fence.

This is not different from regular write requests. What we are going to do is to
1. try to drain ongoing requests on all nodes (if some node fails this it is not a critical problem)
2. erect fence on affected nodes (those that are replicas for the requests)  (if this fails we may fall back to waiting or cancel the op if possible)

After step 2 if some coordinators are on the old topology they will not be able to do any requests until they catch up.

Reply:
> After step 2 if some coordinators are on the old topology they will not be able to do any requests until they catch up.

They will be able to complete requests using themselves as replicas and other replicas that also didn't get the new topology.

Reply:
> > After step 2 if some coordinators are on the old topology they will not be able to do any requests until they catch up.
> 
> They will be able to complete requests using themselves as replicas and other replicas that also didn't get the new topology.

Only for rages that were unaffected which is fine. For affected ranges those coordinator are not replicas and hence will not be able to complete requests by themselves.